README.md

```diff
@@ -18,36 +18,203 @@ testing first appeard in Erlang's proprietary QuickCheck. The
+transition m         (Read  _)     _   = m
+transition (Model m) (Write ref i) _   = Model (update ref i         m)
+transition (Model m) (Inc   ref)   _   = Model (update ref (old + 1) m)
+  where
+  Just old       = lookup ref m
+
+update :: Eq a => a -> b -> [(a, b)] -> [(a, b)]
+update ref i m = (ref, i) : filter ((/= ref) . fst) m
+```
+
+Post-conditions are checked after we executed an action and got access to the
+result.
+
+```haskell
+postcondition :: Model Concrete -> Action Concrete resp -> resp -> Property
+postcondition _         New         _    = property True
+postcondition (Model m) (Read ref)  resp = lookup ref m === Just resp
+postcondition _         (Write _ _) _    = property True
+postcondition _         (Inc _)     _    = property True
+```
+
+Finally, we have to explain how to generate and shrink actions.
+
+```haskell
+generator :: Model Symbolic -> Gen (Untyped Action)
+generator (Model m)
+  | null m    = pure (Untyped New)
+  | otherwise = frequency
+      [ (1, pure (Untyped New))
+      , (8, Untyped .    Read  <$> elements (map fst m))
+      , (8, Untyped <$> (Write <$> elements (map fst m) <*> arbitrary))
+      , (8, Untyped .    Inc   <$> elements (map fst m))
+      ]
+
+shrinker :: Action v resp -> [Action v resp]
+shrinker (Write ref i) = [ Write ref i' | i' <- shrink i ]
+shrinker _             = []
+```
 
-┌──────────────────────┐
-│ New                  │
-│                 ⟶ $0 │
-└──────────────────────┘
-            │ ┌────────┐
-            │ │ Inc $0 │
-┌─────────┐ │ │        │
-│ Inc $0  │ │ │        │
-│         │ │ │   ⟶ () │
-│         │ │ └────────┘
-│    ⟶ () │ │
-└─────────┘ │
-┌─────────┐ │
-│ Read $0 │ │
-│     ⟶ 1 │ │
-└─────────┘ │
+We can now define a sequential and a parallel property as follows.
+
+```haskell
+prop_references :: Problem -> Property
+prop_references prb = forAllProgram
+  generator
+  shrinker
+  precondition
+  transition
+  initModel $ \prog ->
+    runAndCheckProgram
+      precondition
+      transition
+      postcondition
+      initModel
+      (semantics prb)
+      ioProperty
+      prog
+
+prop_referencesParallel :: Problem -> Property
+prop_referencesParallel prb = forAllParallelProgram
+  generator
+  shrinker
+  precondition
+  transition
+  initModel $ \parallel ->
+    runParallelProgram (semantics prb) parallel $ \hist ->
+      checkParallelProgram
+        transition
+        postcondition
+        initModel
+        parallel
+        hist
+```
 
+If we run the sequential property without introducing any problems to the
+semantics function, i.e. `quickCheck (prop_references None)`, then the property
+passes. If we however introduce the bug problem, then it will fail with the
+minimal counterexample:
 
 ```
+> quickCheck (prop_references Bug)
+*** Failed! Falsifiable (after 16 tests and 4 shrinks):
+[New (Var 0),Write (Var 0) 5 (Var 2),Read (Var 0) (Var 3)]
+Just 5 /= Just 6
+```
+
```

Comment:
Make a remark why this went wrong, the fact that we did a `Write` in the `[5 .. 10]` range which is "buggy".